flink/src/main/java/org/apache/iceberg/flink/source/StreamingMonitorFunction.java

```diff
@@ -0,0 +1,172 @@
+    this.tableLoader = tableLoader;
+    this.projectedSchema = projectedSchema;
+    this.filterExpressions = filterExpressions;
+    this.options = options;
+  }
+
+  @Override
+  public void initializeState(FunctionInitializationContext context) throws Exception {
+    snapshotIdState = context.getOperatorStateStore().getListState(
+        new ListStateDescriptor<>(
+            "snapshot-id-state",
+            LongSerializer.INSTANCE));
+    tableLoader.open();
+    table = tableLoader.loadTable();
+    if (context.isRestored()) {
+      LOG.info("Restoring state for the {}.", getClass().getSimpleName());
+      startSnapshotId = snapshotIdState.get().iterator().next();
+    } else {
+      Long optionStartSnapshot = options.startSnapshotId();
+      if (optionStartSnapshot != null) {
+        if (!SnapshotUtil.ancestorOf(table, table.currentSnapshot().snapshotId(), optionStartSnapshot)) {
+          throw new IllegalStateException("The option start-snapshot-id " + optionStartSnapshot +
+              " is not an ancestor of the current snapshot");
+        }
+
+        startSnapshotId = optionStartSnapshot;
+      } else {
+        startSnapshotId = DUMMY_START_SNAPSHOT_ID;
+      }
+    }
+  }
+
+  @Override
+  public void run(SourceContext<FlinkInputSplit> ctx) throws Exception {
+    checkpointLock = ctx.getCheckpointLock();
+    while (isRunning) {
+      synchronized (checkpointLock) {
+        monitorAndForwardSplits(ctx);
+      }
+      Thread.sleep(options.monitorInterval().toMillis());
+    }
+  }
+
+  private void monitorAndForwardSplits(SourceContext<FlinkInputSplit> ctx) {
+    table.refresh();
+    Snapshot snapshot = table.currentSnapshot();
+    if (snapshot != null && snapshot.snapshotId() != startSnapshotId) {
+      long snapshotId = snapshot.snapshotId();
+      // Read current static table if startSnapshotId not set.
+      ScanContext newOptions = startSnapshotId == DUMMY_START_SNAPSHOT_ID ?
+          options.copyWithSnapshotId(snapshotId) :
+          options.copyWithAppendsBetween(startSnapshotId, snapshotId);
+
+      FlinkInputSplit[] splits = FlinkSplitGenerator.createInputSplits(table, options);
+      for (FlinkInputSplit split : splits) {
+        ctx.collect(split);
+      }
+      startSnapshotId = snapshotId;
+    }
+  }
+
+  @Override
+  public void cancel() {
+    // this is to cover the case where cancel() is called before the run()
+    if (checkpointLock != null) {
+      synchronized (checkpointLock) {
+        isRunning = false;
+      }
+    } else {
+      isRunning = false;
+    }
+
+    if (tableLoader != null) {
+      try {
+        tableLoader.close();
+      } catch (IOException e) {
+        throw new UncheckedIOException(e);
+      }
+    }
+  }
+
+  @Override
+  public void close() {
+    cancel();
+  }
+
+  @Override
+  public void snapshotState(FunctionSnapshotContext context) throws Exception {
+    snapshotIdState.clear();
+    snapshotIdState.add(startSnapshotId);
```

Comment:
Okay, so it looks like state is always tracked at the snapshot granularity, and the `checkpointLock` is used to guarantee that a snapshot's data is fully emitted or not emitted at all. Is that right?